ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,305 +2,354 @@
 
-  // Contruct a new prognoser using the prognoser factory. The prognoser
-  // will automatically construct an appropriate model, observer and predictor
-  // based on the values specified in the config.
-  m_prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
+  // This rate object is used to sync the cycles up to the provided Hz.
+  // NOTE: If the cycle takes longer than the provided Hz, nothing bad will
+  //       necessarily occur, but the simulation will be out of sync with real
+  //       time.
+  ros::Rate rate(m_gsap_rate_hz);
 
-  if (m_prognoser == nullptr)
+  // Loop through the PowerModelHandlers to update their values and send them to
+  // the bus each loop to trigger predictions.
+  // This loop should run at 0.5Hz, publishing predictions every 2 seconds via rostopics.
+  int start_loops = NUM_MODELS;
+  while(ros::ok())
   {
-    return false;
-  }
-
-  // Initialize the GSAP prognoser
-  auto init_data = composePrognoserData(m_initial_power, m_initial_voltage, m_initial_temperature);
-  m_prognoser->step(init_data);
+    ros::spinOnce();
 
-  m_init_time = system_clock::now();
+    // Set up value modifiers in each model handler for injection later.
+    injectFaults();
 
-  return true;
-}
+    // DEBUG PRINT
+    if (m_timestamp_print_debug)
+    {
+      printTimestamp(m_power_models[0].model_info.timestamp);
+    }
 
-bool PowerSystemNode::initTopics()
-{
-  // Construct the PowerSystemNode publishers
-  m_mechanical_power_raw_pub = m_nh.advertise<Float64>("mechanical_power/raw", 1);
-  m_mechanical_power_avg_pub = m_nh.advertise<Float64>("mechanical_power/average", 1);
-  m_battery_state_of_charge_pub = m_nh.advertise<owl_msgs::BatteryStateOfCharge>("/battery_state_of_charge", 1);
-  m_battery_remaining_useful_life_pub = m_nh.advertise<owl_msgs::BatteryRemainingUsefulLife>("/battery_remaining_useful_life", 1);
-  m_battery_temperature_pub = m_nh.advertise<owl_msgs::BatteryTemperature>("/battery_temperature", 1);
-  // Finally subscribe to the joint_states to estimate the mechanical power
-  m_joint_states_sub = m_nh.subscribe("/joint_states", 1, &PowerSystemNode::jointStatesCb, this);
-  return true;
-}
+    double excessive_draw = 0.0;
 
-void PowerSystemNode::jointStatesCb(const sensor_msgs::JointStateConstPtr& msg)
-{
-  auto power_watts = 0.0;  // This includes the arm + antenna
-  for (auto i = 0; i < ow_lander::NUM_JOINTS; ++i)
-    power_watts += fabs(msg->velocity[i] * msg->effort[i]);
+    // Cycle the power stats in the models.
+    for (int i = 0; i < NUM_MODELS; i++)
+    {
+      double temp_draw = m_power_models[i].model.RunOnce();
+      // Signal the warning message to be displayed after the loop.
+      if (temp_draw > 0)
+      {
+        // Current implementation means if one input exceeds the draw limit,
+        // all inputs will exceed at the same amount, so it doesn't matter which
+        // cell input we take.
+        excessive_draw = temp_draw;
+      }
```

Comment:
This block is a bit confusing.   Would help to explain what the return value of runOnce() means.  I think the logic could also be accomplished with just one of `temp_draw` and `excessive_draw`.

Reply:
The ``temp_draw`` functionality is mostly a holdover at this point from previously when I had a more involved warning message when a node had excessive power draw applied to it. I think I'll rework this back to what it was before and see about implementing a simpler check for it excessive power draw was calculated.